javascript/ql/src/semmle/javascript/frameworks/Electron.qll

```diff
@@ -6,31 +6,444 @@ module Electron {
+      }
+    }
+    
+    /**
+     * A data flow node that is a call to a synchronous Electron IPC method in the renderer process.
+     */
+    class RendererSendSync extends DataFlow::MethodCallNode {
+      DataFlow::Node channel;
+      
+      RendererSendSync() {
+        exists(Renderer ipcRenderer |
+          this = ipcRenderer.getAMemberCall("sendSync") and
+          channel = this.getArgument(0)
+        )
+      }
+      
+      DataFlow::Node getChannel() {
+        result = channel
+      }
+      
+      string getChannelValue() {
+        result = channel.asExpr().getStringValue()
+      }
+    }
+  
+    
+    /**
+     * A data flow node that is sent as an asynchronous Electron IPC message from the main process via a `webContents` object.
+     */
+    class WebContentsSendMessage extends MainMessage {
+      DataFlow::Node channel;
+      
+      WebContentsSendMessage() {
+        exists(WebContents wc, DataFlow::MethodCallNode mc |
+          wc.flowsTo(mc.getReceiver()) and
+          this = mc.getArgument(1) and
+          channel = mc.getArgument(0) and
+          mc.getCalleeName() = "send"
+        )
+      }
+      
+      override DataFlow::Node getChannel() {
+        result = channel
+      }
+      
+      override string getChannelValue() {
+        result = channel.asExpr().getStringValue()
+      }
+    }
+    /**
+     * Holds if `pred` flows to `succ` via the Electron IPC channel `channel`, as determined from local string values.
+     */
+    predicate simpleFlowStep(DataFlow::Node pred, DataFlow::Node succ, string channel) {
+      // match a message sent from the main thread with a callback parameter in the renderer thread
+      exists(RendererCallback callback |
+        succ = callback.getParameter(1) and
+        channel = callback.getChannelValue() and
+        channel = pred.(MainMessage).getChannelValue()
+      )
+      or
+      // match a message sent from the renderer thread with a callback parameter in the main thread
+      exists(MainCallback callback |
+        succ = callback.getParameter(1) and
+        channel = callback.getChannelValue() and
+        channel = pred.(RendererMessage).getChannelValue()
+      )
+      or
+      // match a synchronous reply sent from the main thread with a `sendSync` call in the renderer thread
+      exists(RendererSendSync sendSync |
+        succ = sendSync and
+        channel = sendSync.getChannelValue() and
+        channel = pred.(MainSyncReplyMessage).getChannelValue()
+      )
+      or
+      // match a synchronous reply sent from the renderer thread with a `sendSync` call in the main thread
+      exists(MainSendSync sendSync |
+        succ = sendSync and
+        channel = sendSync.getChannelValue() and
+        channel = pred.(RendererSyncReplyMessage).getChannelValue()
+      )
+    }
+    
+    /**
+     * An additional flow step  via an Electron IPC message.
+     */
+    class IPCAdditionalFlowStep extends DataFlow::AdditionalFlowStep {
```

Comment:
I think this can be `private`, since we are mostly interested in it in terms of its contribution to `AdditionalFlowStep`.